packages/core/src/synthesis/Synthesizer.ts

```diff
@@ -375,238 +352,308 @@ export class Synthesizer {
-          }
-          break;
+            let newStmts: WithStmts<SubStmt>;
+            if (pick.tag === "PredicateDecl") {
+              newStmts = this.generatePredicate(pick, ctx);
+            } else if (pick.tag === "FunctionDecl") {
+              newStmts = this.generateFunction(pick, ctx);
+            } else {
+              newStmts = this.generateConstructor(pick, ctx);
+            }
+            const { res, stmts } = newStmts;
+            let toDelete: SubStmt[];
+            // remove old statement
+            if (
+              res.tag === "Bind" &&
+              res.variable.type !== oldStmt.variable.type
+            ) {
+              // old bind was replaced by a bind with diff type
+              toDelete = cascadingDelete(oldStmt, this.currentProg); // remove refs to the old bind
+            } else {
+              toDelete = [oldStmt];
+            }
+            const deleteOps: Delete[] = toDelete.map((s) => ({
+              tag: "Delete",
+              stmt: s,
+              mutate: removeStmtCtx,
+            }));
+            const addOps: Add[] = stmts.map((stmt) => ({
+              tag: "Add",
+              stmt,
+              mutate: appendStmtCtx,
+            }));
+            return {
+              newStmt: res,
+              additionalMutations: [...deleteOps, ...addOps],
+            };
+          } else return undefined;
         }
-      }
-    } else {
-      log.debug(
-        `Failed to find a statement when editing. Candidates are: ${candidates}. Chosen name to find is ${chosenName} of ${chosenType}`
-      );
-    }
+      ),
+    ];
+    const mutations = compact(ops);
+    log.debug(
+      `Available mutations for ${prettyStmt(stmt)}:\n${showOps(mutations)}`
+    );
+    return mutations.map((m) => [m]);
   };
 
-  typeChange = (oldStmt: ApplyPredicate | Bind, pick: ArgStmtDecl): void => {
-    let newStmt = oldStmt;
-    if (pick.tag === "PredicateDecl") {
-      newStmt = this.generatePredicate(pick);
-    } else if (pick.tag === "FunctionDecl") {
-      newStmt = this.generateFunction(pick);
-    } else {
-      newStmt = this.generateConstructor(pick);
-    }
-    // remove old statement
-    if (
-      newStmt.tag === "Bind" &&
-      oldStmt.tag === "Bind" &&
-      newStmt.variable.type !== oldStmt.variable.type
-    ) {
-      // old bind was replaced by a bind with diff type
-      this.cascadingDelete(oldStmt); // remove refs to the old bind
-      newStmt = newStmt as Bind;
-    } else {
-      // otherwise we can simple delete
-      this.cxt.removeStmt(oldStmt);
-    }
-    this.cxt.ops.push({
-      tag: "Replace",
-      old: oldStmt,
-      new: newStmt,
-      mutationType: "TypeChange",
-    });
+  /**
+   * Pick a random statement in the Substance program and enumerate all the applicable mutations.
+   * @returns a list of mutation groups, each representing a series of `Update` mutations
+   */
+  editStmt = (ctx: SynthesisContext): MutationGroup[] => {
```

Comment:
nit: potentially consider renaming this function as well as the `addStmt` and `deleteStmt` functions to clarify that they are generating the list of all possible mutations for the chosen statement, not making any changes to the AST. Something like `editCandidates` or `buildEditGroup` or `buildEditCandidates`, etc.

Reply:
Yeah, they are really enumerating things. Changed the names to reflect this fact.